### ISSUE TITLE
In install.py consider using site-packages as bindings-dir to avoid the need of PYTHONPATH extension

### ISSUE THREAD

**Issue by mikand** [pysmt/pysmt#495]
This point was raised by @randomir in #494, and I think it could make a lot of sense.

Let's discuss a bit the pros and cons before taking a decision.

**Comment by randomir** [comment]
@marcogario I like your proposal to just copy bindings if they already exist in the install dir. The only problem I see with that approach is making sure the compiled binaries are up to date -- consider user installs an old version of `pysmt`, installs solvers (does compilation), then upgrades `pysmt` (or installs new version in new venv), installs solvers, and gets the old binaries.

On your point of using local packages directory -- it seems doable.

So we would want:
- if venv is active, install in venv's `site-packages`
- otherwise:
  - is user has local site packages enabled (`site.ENABLE_USER_SITE`), install in user site (`~/.local/...`)
  - otherwise attempt global install (`/usr/lib/python/...`)

Right?

**Comment by marcogario** [comment]
Do we need to detect the venv? Wouldn't the installation at site packages be system wide if a venv is not in use? I am just concerned that detection of the venv might be a pain. For the rest I agree on the approach.

Regarding stale bindings, we do not have this problem now, and I think we should not have it if we have venvs. Moving the bindings forces a recompilation. I would keep this as it is,and try to optimize if we really need to.

**Comment by randomir** [comment]
I hope we don’t need to detect venv, I wasn’t trying to say we do that — my outline above was more of functional specification, than actual implementation steps. :)

Ok, I’ll play around with it a little bit.

**Comment by marcogario** [comment]
Any progress on this? I think this would be an important improvement on usability. Let me know if I can help.

**Comment by randomir** [comment]
You are right, sorry, I’ve been busy with other stuff. I’ll get back to this, hopefully have a PR soon.

Btw, unrelated, when do you plan your next release?

**Comment by marcogario** [comment]
No worries, just let me know if we can help.

We do not have a strict schedule for releases, but we try to make 2 releases a year. Recently we got a few bugfixes that I think would justify making a release of 0.7.6 by the end of July.

**Comment by randomir** [comment]
@marcogario, so I guess my question is, do we want to explore this idea further (if you already haven't), or to proceed with my original proposal**?

** which, btw, mutated to the following (and I'm not sure we can do better):
- if user has virtual environment activated, simply install solvers in virtualenv's site packages
- otherwise (debian's pip logic):
  - if user is root, install to system site/dist packages
  - else if user has user site enabled, install solvers user-local
  - otherwise, try installing to system site packages and probably fail.

The implementation is simple, using `distutils`, and not relying on manual testing of each branch above.

**Comment by marcogario** [comment]
I would move forward with the original proposal. 

Packaging would definitely be a better solute, but there are several issues,mostly related to effort. To truly create packages that do not require compilation, you end up with a huge matrix of combinations. For example, on osx you might install python in different ways (conda, homebrew). The version of python that you use needs to match (to some extent). This leads to #solvers x #pythonVersions x #os x #pythonBinary number of packages to maintain (that's at least 70 packages). 

The other problem is licensing. I believe now we have only a restriction on Mathsat, but when we started most solvers where not opensource. This might be easier to adress now. 

Finally, the current install.py enforces the version of the solvers without requiring external dependency mgmt. We currently do not support multiple releases of the solvers within the same version of pysmt. This simplifies significantly the release process and testing. 

However, if we start by installing the bindings in a reasonable place, this would simplify moving to a package oriented version of install.py. I think this was what we experimented doing with picosat. Meaning that whether you installed tge solver with pip or install.py, it shouldn't matter. 

What do you think?

**Comment by randomir** [comment]
Sure, packaging solvers for pypi requires substantial effort (even z3 community isn't willing to close that, https://github.com/Z3Prover/z3/issues/288). But there's a middle ground - providing wheels for only some combinations in that matrix and compiling from source on pip install for others. Details depend on solver. E.g. you currently install z3 from 3rd-party archive, and they provide only ~10 combinations of platform/python. Cvc4 you always compile on install (so it seems).

A major concern I have with this simple copying of bindings to site-packages is that these bindings only partially act as python packages -- since you can import them, but you can't uninstall them, list with pip, or check versions. Because they are not actual packages.

Current approach with custom bindings path, and required manual user intervention, is inconvenient, but at least explicit.

That being said, the proposed approach probably does improve experience for an average user, the caveat being non-standard "install" is now less explicit. We should probably clarify that in `pysmt-install`. OTOH, user still needs to "opt-in" by manually running the non-standard installer.

So, long story short, I agree with proceeding, just wanted to lay out and acknowledge caveats before doing that.

**Comment by marcogario** [comment]
You raise a fair point with the "uninstall" issue. We did not pay much attention to this, but this could be a "good citizen" point. I think we should move ahead as discussed.

 In parallel, if you want to make a demo of how the plugin approach would work, that would be great, so that we can discuss pros/cons concretely. For the foreseeable future, I would keep pysmt-install as front-end (since this also provides some services for CI to check that the solver binding is *really* working) but we could slowly package things up and iron out issues as we face them.

**Comment by randomir** [comment]
Completely agree. `pysmt-install` can exist in parallel, as packages for solvers are created. It can even use the package instead of the custom installer if package is available for the particular solver.

After finishing this PR, I'll proceed with a demo package for a solver.

**Comment by randomir** [comment]
The #517 implements the logic we discussed, but detecting virtual env was unfortunately unavoidable. Namely, we need to determine somehow the case in which we're doing the `--user` install -- and that's when we're running outside of venv.

Travis tests verify this is working for the existing matrix of environments. My only concern are exotic envs like conda on windows.

**Comment by marcogario** [comment]
See #517 .
